File: 07-security/03-xss-and-output-context/golang/xss_context.go
// Layer 7 · Topic 3 — XSS by output context (Go / html/template).
//
// One command, no arguments: `go run xss_context.go`.
// Go's html/template is the ONE template engine here that actually solves the
// context problem: it PARSES the template as HTML, tracks which context each
// {{.}} sits in (body, attribute, href/URL, inside <script>), and picks a
// different escaper per context. Put a javascript: URL in an href and it
// emits #ZgotmplZ instead of the URL. This program renders the SAME attacker
// payload into four contexts and prints the exact bytes html/template emits,
// so you can read, per context, whether the dangerous construct survived.
//
// The determination is made from the emitted bytes, not a browser run: a cell
// is "EXECUTES" only if a live <script>, an on* handler, or a javascript:
// scheme reached the output intact. Compare to Topic 2's parameterization --
// it is the same move: make the tool understand the output language instead of
// treating it as opaque text.
//
// What to look for: every context is neutralized. html/template makes the
// context problem disappear rather than documenting it.
package main

import (
	"bytes"
	"fmt"
	"html/template"
	"strings"
)

type cell struct {
	name    string
	tmpl    string
	payload string
}

// A crude but honest "would this run in a browser?" check over emitted bytes.
func executes(html string) bool {
	low := strings.ToLower(html)
	// A construct executes only if it reached the output UNESCAPED. Each test
	// requires the literal active bytes; an entity-escaped version (&#34;,
	// &lt;) does not match, which is exactly the neutralization we are checking.
	switch {
	case strings.Contains(low, "<script>beacon()</script>"):
		return true // a live script element (body or break-out)
	case strings.Contains(low, "\" onmouseover=beacon()"):
		return true // the attribute quote survived -> handler opened
	case strings.Contains(low, "href=\"javascript:beacon()") ||
		strings.Contains(low, "href='javascript:beacon()"):
		return true // a live javascript: scheme in an href
	}
	return false
}

func main() {
	fmt.Println("Layer 7 · Topic 3 — XSS by output context (Go / html/template)\n")
	cells := []cell{
		{"body-context <script>", `<div>{{.}}</div>`, `<script>beacon()</script>`},
		{"attribute handler", `<a title="{{.}}">x</a>`, `" onmouseover=beacon() x="`},
		{"javascript: URL", `<a href="{{.}}">x</a>`, `javascript:beacon()`},
		{"</script> break-out", `<script>var c = "{{.}}";</script>`, `</script><script>beacon()</script>`},
	}
	fmt.Printf("  %-24s %-9s emitted bytes\n", "context", "executes?")
	for _, c := range cells {
		t := template.Must(template.New(c.name).Parse(c.tmpl))
		var buf bytes.Buffer
		_ = t.Execute(&buf, c.payload)
		out := buf.String()
		ex := "NO"
		if executes(out) {
			ex = "YES"
		}
		fmt.Printf("  %-24s %-9s %s\n", c.name, ex, out)
	}
	fmt.Println("\nRead: the javascript: URL becomes #ZgotmplZ, the <script> payload is\n" +
		"entity-escaped in body AND JS-string-escaped inside <script>, and the\n" +
		"attribute quote is escaped so the handler never opens. Correct-by-\n" +
		"construction, because the engine understands the output language.")
	fmt.Println("\nThe cost (README Q4): html/template can MANGLE a legitimate template\n" +
		"it cannot prove safe -- e.g. building a URL from a trusted constant plus a\n" +
		"value can emit #ZgotmplZ where you wanted a real link. Jinja never does\n" +
		"that because Jinja never looked at the context at all.")
}

File: 07-security/03-xss-and-output-context/python/xss_context.py
"""
Layer 7 · Topic 3 — XSS by output context (Python / Jinja2).

One command, no arguments: `python3 xss_context.py`.
Jinja2's autoescape escapes for ONE context -- HTML body -- and does not
parse the surrounding HTML, so it cannot know where a value landed. This
program renders the SAME attacker payload into four contexts, twice (with
autoescape ON, the FastAPI/`.html` default, and OFF, the bare Environment()
trap), and prints the exact emitted bytes plus whether the payload would
execute in a browser.

The execute verdict is a STATIC determination from the emitted bytes (the
full lab fires these in a real headless browser and counts beacon bytes):
body/attribute/href are adjudicated by parsing the output with stdlib
html.parser -- so an entity-escaped &#34; correctly reads as text, not an
attribute boundary -- and the <script> break-out is judged by whether a
literal </script> was injected.

What to look for: with autoescape ON, body and the two break-outs are
neutralized, but the `javascript:` URL still EXECUTES -- Jinja escaped the
HTML metacharacters (there are none in `javascript:beacon()`) and has no idea
`href` is a URL context. That single surviving cell is the whole lesson, and
it is exactly the cell Go's html/template turns into #ZgotmplZ.
"""
from html.parser import HTMLParser
from jinja2 import Environment


class TagScan(HTMLParser):
    """Collects (tag, attrs-dict) and script data from emitted HTML.
    Attribute values arrive already entity-decoded, and <script> content is
    read in CDATA mode -- both match how a browser tokenizes."""
    def __init__(self):
        super().__init__(convert_charrefs=True)
        self.tags = []
        self.script_data = []
        self._in_script = False

    def handle_starttag(self, tag, attrs):
        self.tags.append((tag, dict(attrs)))
        if tag == "script":
            self._in_script = True
            self.set_cdata_mode("script")

    def handle_endtag(self, tag):
        if tag == "script":
            self._in_script = False

    def handle_data(self, data):
        if self._in_script:
            self.script_data.append(data)


def executes(context, emitted):
    scan = TagScan()
    scan.feed(emitted)
    if context == "body":
        # A live <script> element in the body executes.
        return "<script>beacon()</script>" in emitted
    if context == "attr":
        # A real event-handler attribute (not text inside a value).
        return any(any(a.startswith("on") and "beacon" in v for a, v in attrs.items())
                   for _tag, attrs in scan.tags)
    if context == "href":
        # An href whose scheme is javascript: after decoding.
        return any(v.strip().lower().startswith("javascript:") and "beacon" in v
                   for _tag, attrs in scan.tags for a, v in attrs.items() if a == "href")
    if context == "script":
        # The injection closed the script and opened a new one.
        return emitted.lower().count("</script>") > 1
    return False


CELLS = [
    ("body",   "body-context <script>", "<div>{{ x }}</div>",              "<script>beacon()</script>"),
    ("attr",   "attribute handler",     "<a title={{ x }}>x</a>",          "y onmouseover=beacon()"),
    ("href",   "javascript: URL",       '<a href="{{ x }}">x</a>',         "javascript:beacon()"),
    ("script", "</script> break-out",   '<script>var c="{{ x }}";</script>', "</script><script>beacon()</script>"),
]


def run(autoescape):
    env = Environment(autoescape=autoescape)
    label = "autoescape ON " if autoescape else "autoescape OFF"
    print(f"  --- {label} (Jinja2) ---")
    print(f"    {'context':<24}{'executes?':<10}emitted bytes")
    executed = 0
    for ctx, name, tmpl, payload in CELLS:
        out = env.from_string(tmpl).render(x=payload)
        ex = executes(ctx, out)
        executed += ex
        print(f"    {name:<24}{('YES' if ex else 'NO'):<10}{out}")
    print(f"    -> {executed}/4 payloads executed\n")


def main():
    print("Layer 7 · Topic 3 — XSS by output context (Python / Jinja2)\n")
    run(autoescape=True)
    run(autoescape=False)
    print("Read: autoescape ON stops the body and both break-outs, because it "
          "escapes <, >, \" and '. It does NOT stop the javascript: URL -- that "
          "payload has no HTML metacharacters, so HTML escaping is a no-op, and "
          "Jinja never modeled that href is a URL context. autoescape OFF (the "
          "bare Environment() default) executes all four. Compare Go's "
          "html/template, which neutralizes the URL cell too.")


if __name__ == "__main__":
    main()
